File: src/sys/pkg/testing/host-target-testing/packages/package.go
// Copyright 2019 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package packages

import (
	"bytes"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"

	"go.fuchsia.dev/fuchsia/garnet/go/src/far"
	"go.fuchsia.dev/fuchsia/src/sys/pkg/bin/pm/build"
)

type FileData []byte

type Package struct {
	merkle   string
	repo     *Repository
	contents build.MetaContents
}

// newPackage extracts out a package from the repository.
func newPackage(repo *Repository, merkle string) (Package, error) {
	// Need to parse out the package meta.far to find the package contents.
	blob, err := repo.OpenBlob(merkle)
	if err != nil {
		return Package{}, err
	}
	defer blob.Close()

	f, err := far.NewReader(blob)
	if err != nil {
		return Package{}, err
	}
	defer f.Close()

	b, err := f.ReadFile("meta/contents")
	if err != nil {
		return Package{}, err
	}

	contents, err := build.ParseMetaContents(bytes.NewReader(b))
	if err != nil {
		return Package{}, err
	}

	return Package{
		merkle:   merkle,
		repo:     repo,
		contents: contents,
	}, nil
}

// Merkle returns the meta.far merkle.
func (p *Package) Merkle() string {
	return p.merkle
}

// Open opens a file in the package.
func (p *Package) Open(path string) (*os.File, error) {
	merkle, ok := p.contents[path]
	if !ok {
		return nil, os.ErrNotExist
	}

	return p.repo.OpenBlob(merkle.String())
}

// ReadFile reads a file from a package.
func (p *Package) ReadFile(path string) ([]byte, error) {
	r, err := p.Open(path)
	if err != nil {
		return nil, err
	}

	return ioutil.ReadAll(r)
}

func (p *Package) Expand(dir string) error {
	for path := range p.contents {
		data, err := p.ReadFile(path)
		if err != nil {
			return fmt.Errorf("invalid path. %w", err)
		}
		realPath := filepath.Join(dir, path)
		if err := os.MkdirAll(filepath.Dir(realPath), 0755); err != nil {
			return fmt.Errorf("could not create parent directories for %s, %w", realPath, err)
		}
		if err = ioutil.WriteFile(realPath, data, 0644); err != nil {
			return fmt.Errorf("could not export %s to %s. %w", path, realPath, err)
		}
	}
	blob, err := p.repo.OpenBlob(p.merkle)
	if err != nil {
		return fmt.Errorf("failed to open meta.far blob. %w", err)
	}
	defer blob.Close()

	f, err := far.NewReader(blob)
	if err != nil {
		return fmt.Errorf("failed to open reader on blob. %w", err)
	}
	defer f.Close()

	for _, path := range f.List() {
		data, err := f.ReadFile(path)
		if err != nil {
			fmt.Errorf("failed to read %s. %w", path, err)
		}
		realPath := filepath.Join(dir, path)
		if err := os.MkdirAll(filepath.Dir(realPath), 0755); err != nil {
			return fmt.Errorf("could not create parent directories for %s, %w", realPath, err)
		}
		if err = ioutil.WriteFile(realPath, data, 0644); err != nil {
			fmt.Errorf("failed to write file %s. %w", realPath, err)
		}
	}

	return nil
}

File: src/security/scrutiny/lib/plugins/src/toolkit/controller/package.rs
// Copyright 2020 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use {
    anyhow::anyhow,
    anyhow::Result,
    fuchsia_archive::Reader as FarReader,
    scrutiny::{
        model::controller::{ConnectionMode, DataController, HintDataType},
        model::model::*,
    },
    scrutiny_utils::{env, usage::*},
    serde::{Deserialize, Serialize},
    serde_json::{json, value::Value},
    std::fs::{self, File},
    std::io::prelude::*,
    std::io::Cursor,
    std::path::PathBuf,
    std::sync::Arc,
};

#[derive(Deserialize, Serialize)]
pub struct PackageExtractRequest {
    // The input path for the ZBI.
    pub url: String,
    // The output directory for the extracted ZBI.
    pub output: String,
}

pub const BLOBS_PATH: &str = "amber-files/repository/blobs";

#[derive(Default)]
pub struct PackageExtractController {}

impl DataController for PackageExtractController {
    fn query(&self, model: Arc<DataModel>, query: Value) -> Result<Value> {
        let fuchsia_build_dir = env::fuchsia_build_dir()?;
        let blob_dir = fuchsia_build_dir.join(BLOBS_PATH);

        let request: PackageExtractRequest = serde_json::from_value(query)?;
        let packages = model.packages().read().unwrap();
        for package in packages.iter() {
            if package.url == request.url {
                let output_path = PathBuf::from(request.output);
                fs::create_dir_all(&output_path)?;

                let pkg_path = blob_dir.clone().join(package.merkle.clone());
                let mut pkg_file = File::open(pkg_path)?;
                let mut pkg_buffer = Vec::new();
                pkg_file.read_to_end(&mut pkg_buffer)?;

                let mut cursor = Cursor::new(pkg_buffer);
                let mut far = FarReader::new(&mut cursor)?;

                let pkg_files: Vec<String> = far.list().map(|s| String::from(s)).collect();
                // Extract all the far meta files.
                for file_name in pkg_files.iter() {
                    let data = far.read_file(file_name)?;
                    let file_path = output_path.clone().join(file_name);
                    if let Some(parent_dir) = file_path.as_path().parent() {
                        fs::create_dir_all(parent_dir)?;
                    }
                    let mut package_file = File::create(file_path)?;
                    package_file.write_all(&data)?;
                }

                // Extract all the contents of the package.
                for (file_name, blob) in package.contents.iter() {
                    let file_path = output_path.clone().join(file_name);

                    let blob_path = blob_dir.clone().join(blob);
                    let mut blob_file = File::open(blob_path)?;
                    let mut blob_buffer = Vec::new();
                    blob_file.read_to_end(&mut blob_buffer)?;

                    if let Some(parent_dir) = file_path.as_path().parent() {
                        fs::create_dir_all(parent_dir)?;
                    }
                    let mut package_file = File::create(file_path)?;
                    package_file.write_all(&blob_buffer)?;
                }
                return Ok(json!({"status": "ok"}));
            }
        }
        Err(anyhow!("Unable to find package url"))
    }

    fn description(&self) -> String {
        "Extracts a package from a url to a directory.".to_string()
    }

    fn usage(&self) -> String {
        UsageBuilder::new()
            .name("tool.package.extract - Extracts Fuchsia package to a directory.")
            .summary("tool.package.extract --url fuchsia-pkg://fuchsia.com/foo --output /tmp/foo")
            .description(
                "Extracts package from a given url to some provided file path. \
                Internally this is resolving the URL and extracting the internal
                Fuchsia Archive and resolving all the merkle paths.
                ",
            )
            .arg("--url", "Package url that you wish to extract.")
            .arg("--output", "Path to the output directory")
            .build()
    }

    fn hints(&self) -> Vec<(String, HintDataType)> {
        vec![
            ("--url".to_string(), HintDataType::NoType),
            ("--output".to_string(), HintDataType::NoType),
        ]
    }

    /// PackageExtract is only available to the local shell as it directly
    /// modifies files on disk.
    fn connection_mode(&self) -> ConnectionMode {
        ConnectionMode::Local
    }
}
